src/java/test/org/apache/zookeeper/server/ServerCnxnTest.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.zookeeper.server;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import org.apache.jute.Record;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.ZKTestCase;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.proto.ReplyHeader;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ServerCnxnTest extends ZKTestCase {
+
+    /**
+     * Test getting a copy of authinfo to avoid parallel modification impact
+     */
+    @Test
+    public void testServerCnxnGetAuthInfoWithCopy() throws Exception {
+        MockServerCnxn serverCnxn = new MockServerCnxn();
+        List<Id> authInfo = serverCnxn.getAuthInfo();
+        Id id = new Id("testscheme", "test");
+        serverCnxn.addAuthInfo(id);
+        Assert.assertFalse(authInfo.contains(id));
+        Assert.assertTrue(serverCnxn.getAuthInfo().contains(id));
+    }
+    
+    /**
+     * Mock extension of ServerCnxn dummy to test for
+     * AuthInfo behavior (ZOOKEEPER-2977).
+     */
+    private static class MockServerCnxn extends ServerCnxn {
```

Comment:
@sumitagrawl Can't you just use mockito to get the desired behaviour?

Reply:
Followed similar test case as MockNIOServerCnxn.java & NIOServerCnxnTest.java as present in same package, so should change this to use Mockito?

Reply:
Up to you. 

But, it's generally better to keep the scope of this patch for changes strictly related to the Jira and create separate one for other refactorings.

I'm happy to commit both, but you need to do the paperwork.

Reply:
Ok, will change to Mockito.
Please share some details related to the paper work to be done.

Reply:
Cool, thanks @sumitagrawl 
Sorry I was trying to kinda joking with "paperwork": you only need to create separate Jira and pull request for the refactoring. ;)